modules/generative-octoai/clients/octoai.go

```diff
@@ -0,0 +1,222 @@
+
+	res, err := v.httpClient.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "send POST request")
+	}
+	defer res.Body.Close()
+
+	bodyBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "read response body")
+	}
+
+	var resBody generateResponse
+	if err := json.Unmarshal(bodyBytes, &resBody); err != nil {
+		return nil, errors.Wrap(err, "unmarshal response body")
+	}
+
+	if res.StatusCode != 200 || resBody.Error != nil {
+		if resBody.Error != nil {
+			return nil, errors.Errorf("connection to OctoAI API failed with status: %d error: %v", res.StatusCode, resBody.Error.Message)
+		}
+		return nil, errors.Errorf("connection to OctoAI API failed with status: %d", res.StatusCode)
+	}
+
+	textResponse := resBody.Choices[0].Message.Content
+
+	return &generativemodels.GenerateResponse{
+		Result: &textResponse,
+	}, nil
+}
+
+func (v *octoai) getOctoAIUrl(ctx context.Context, baseURL string) (string, error) {
+	passedBaseURL := baseURL
+	if headerBaseURL := v.getValueFromContext(ctx, "X-Octoai-Baseurl"); headerBaseURL != "" {
+		passedBaseURL = headerBaseURL
+	}
+	return url.JoinPath(passedBaseURL, "/v1/chat/completions")
+}
+
+func (v *octoai) generatePromptForTask(textProperties []map[string]string, task string) (string, error) {
+	marshal, err := json.Marshal(textProperties)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf(`'%v:
+%v`, task, string(marshal)), nil
+}
+
+func (v *octoai) generateForPrompt(textProperties map[string]string, prompt string) (string, error) {
+	all := compile.FindAll([]byte(prompt), -1)
+	for _, match := range all {
+		originalProperty := string(match)
+		replacedProperty := compile.FindStringSubmatch(originalProperty)[1]
+		replacedProperty = strings.TrimSpace(replacedProperty)
+		value := textProperties[replacedProperty]
+		if value == "" {
+			return "", errors.Errorf("Following property has empty value: '%v'. Make sure you spell the property name correctly, verify that the property exists and has a value", replacedProperty)
+		}
+		prompt = strings.ReplaceAll(prompt, originalProperty, value)
+	}
+	return prompt, nil
+}
+
+func (v *octoai) getValueFromContext(ctx context.Context, key string) string {
+	if value := ctx.Value(key); value != nil {
+		if keyHeader, ok := value.([]string); ok && len(keyHeader) > 0 && len(keyHeader[0]) > 0 {
+			return keyHeader[0]
+		}
+	}
+	// try getting header from GRPC if not successful
+	if apiKey := modulecomponents.GetValueFromGRPC(ctx, key); len(apiKey) > 0 && len(apiKey[0]) > 0 {
+		return apiKey[0]
+	}
+	return ""
+}
+
+func (v *octoai) getApiKey(ctx context.Context) (string, error) {
+	if apiKey := v.getValueFromContext(ctx, "X-Octoai-Api-Key"); apiKey != "" {
+		return apiKey, nil
+	}
+	if v.apiKey != "" {
+		return v.apiKey, nil
+	}
+	return "", errors.New("no api key found " +
+		"neither in request header: X-Octoai-Api-Key " +
+		"nor in environment variable under OPENAI_API_KEY")
```

Comment:
You should change `X-Octoai-Api-Key` to `X-OctoAI-Api-Key` and `OPENAI_API_KEY` to `OCTOAI_APIKEY`

Reply:
Ack on `X-OctoAI-Api-Key`

And whoops that should have been changed from OPENAI -> OCTOAI for sure. That said, we use `OCTOAI_API_KEY`instead of `OCTOAI_APIKEY` a lot in our documentation and code examples (e.g. https://octo.ai/docs/text-gen-solution/using-unstructured-io-for-embedding-documents). I'm happy to use the prescribed formatting for it to work best for Weaviate users but I figured I'd point this out.